pyquil/analytics.py

```diff
@@ -0,0 +1,310 @@
+#!/usr/bin/python
+##############################################################################
+# Copyright 2016-2017 Rigetti Computing
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+"""
+A collection of methods for analyzing pyquil programs (gate count, etc...)
+"""
+
+import pyquil.gates as gates
+from pyquil.quilbase import UnaryClassicalInstruction, BinaryClassicalInstruction, SimpleInstruction, QuilAction
+from grove.qft.fourier import qft
+from grove.unitary_circuit.arbitrary_state import create_arbitrary_state
+import numpy as np
+from pyquil.quil import Program
+from grove.teleport.teleportation import make_bell_pair
+import plotly
+import plotly.graph_objs as go
+from copy import deepcopy
+import time
+import random
+
+def get_total_instruction_count(program):
+    '''
+    Returns the total number of instructions within the given program
+    :param program: The program to analyze
+    :return: The max number of instructions executed by this program
+    '''
+    
+    return len(program.actions)
+
+def get_qubit_count(program):
+    '''
+    Returns the total number of qubits needed for this program
+    :param program: The program to analyze
+    :return: The total number of qubits allocated for this program
+    '''
+    return len(program.get_qubits())
+
+def get_gate_types(program):
+    '''
+    Returns a dictionary of gate types to their counts within the program
+    :param program: The program to analyze
+    :return: The mapping of gates to their counts within program
+    '''
+    counts = {}
+    for gate in get_all_gate_types():
+        counts[gate] = 0
+    
+    # Count how many gates there are for each gate type
+    for action in program.actions:
+        try:
+            counts[action[1].operator_name] += 1
+        except Exception:
+            try:
+                counts[action[1].op] += 1
+            except:
+                pass#print action[1]
+        
+    return counts
+
+def get_qubit_num_gates(gate_counts):
+    '''
+    Returns a dictionary with the original gate_counts, but with
+    extra entries 'Single Qubit Gates', 'Two Qubit Gates',
+    'Control Instructions', 'Classical Instructions',
+    'Three+ Qubit Gates', and 'Elements in DEFGATES'
+    '''
+    
+    results = deepcopy(gate_counts)
+    results['Single Qubit Gates'] = \
```

Comment:
this is scary looking!!  Better would be to have these categories constructed more dynamically, e.g. they should be lists of instruction objects not strings.  As an example, better would be to have a list
```
from pyquil.gates import *
single_q_gates = [I, H, ...]
```
where I and H are the actual gate objects not just the strings.  This is more flexible.  The idea is to (as much as possible) have a single source of truth in the pyquil library for where gates are defined and mapped to their string representations. Does that make sense?

Then you can just iterate through the categories to assemble your dictionary.